File: Association Slope/Linked List/Q3. Reverse Linked List/Q3_ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // Base case: if head is null or only one node, it's the new head
        if (head == null || head.next == null) {
            return head;
        }

        // Recursively reverse the rest of the list
        ListNode reversedHead = reverseList(head.next);

        // Reverse current node's pointer
        head.next.next = head;  // Make the next node point to current
        head.next = null;       // Break the current link

        return reversedHead;    // Return new head node
    }
}

File: Association Slope/Linked List/Q3. Reverse Linked List/Q3_ReverseLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val          # Node value
#         self.next = next        # Reference to next node

class Solution(object):
    def reverseList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # Base case: if list is empty or has one node, it's already reversed
        if head is None or head.next is None:
            return head

        # Recursively reverse the rest of the list
        new_head = self.reverseList(head.next)

        # Reverse current node's pointer
        head.next.next = head    # Make next node point back to current
        head.next = None         # Break original forward link

        return new_head          # Return new head of the reversed list
